### PR TITLE
[Backport 2.x] Integration tests framework

### PR DESCRIPTION
**Pull request by peternied**
### Description
Test framework that allows for declaring within the test class all
related users, roles, and other security primitives.  Emphasis on
hamcrest Matchers for straight forward to read, author, and
troubleshoot test cases.

#### Special Note
This is a _giant_ pull request - I would highly recommend using the github.dev [[view]](https://github.com/opensearch-project/security/pull/3388) for reviewing and providing feedback.  I'll also provide comments for the initial reviewers where I call out any deviations I've taken from what is in main

### Issues Resolved
- Resolves https://github.com/opensearch-project/security/issues/3230

### Check List
- [X] New functionality includes testing
- [ ] New functionality has been documented
- [X] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by peternied** [comment]
Tests with failures  - tracking burndown:
 - [ ] org.opensearch.security.DefaultConfigurationTests.shouldLoadDefaultConfiguration
 - [ ] org.opensearch.security.PointInTimeOperationTest.listPitSegmentsCreatedWithIndexAlias_positive
 - [ ] org.opensearch.security.PointInTimeOperationTest.listPitSegments_negative
 - [ ] org.opensearch.security.PointInTimeOperationTest.listPitSegments_positive
 - [ ] org.opensearch.security.PointInTimeOperationTest.listPitSegmentsCreatedWithIndexAlias_negative
 - [ ] org.opensearch.security.api.DashboardsInfoTest.testDashboardsInfoValidationMessage
 - [ ] org.opensearch.security.api.DashboardsInfoWithSettingsTest.testDashboardsInfoValidationMessageWithCustomMessage
 - [ ] org.opensearch.security.rest.AuthZinRestLayerTests.testAccessDeniedForUserWithNoPermissions
 - [ ] org.opensearch.security.rest.AuthZinRestLayerTests.testShouldAllowAtRestAndTransport
 - [ ] org.opensearch.security.rest.AuthZinRestLayerTests.testBackwardsCompatibility
 - [ ] org.opensearch.security.rest.AuthZinRestLayerTests.testShouldAllowAtRestAndBlockAtTransport
 - [ ] org.opensearch.security.rest.AuthZinRestLayerTests.testShouldBlockAccessToEndpointForWhichUserHasNoPermission
 - [ ] org.opensearch.security.rest.WhoAmITests.testWhoAmIWithGetPermissionsLegacy
 - [ ] org.opensearch.security.rest.WhoAmITests.testWhoAmIWithoutGetPermissions
 - [ ] org.opensearch.security.rest.WhoAmITests.testWhoAmIWithGetPermissions
 - [ ] org.opensearch.security.rest.WhoAmITests.testAuditLogSimilarityWithTransportLayer

File: .github/workflows/ci.yml
Comment (peternied) [inline .github/workflows/ci.yml:84]: TODO: Merge this into the rest of the dynamically listed tests?

File: .github/workflows/ci.yml
Comment (peternied) [inline .github/workflows/ci.yml:1]: Merged in a couple of missed changes from main

File: build.gradle
Comment (peternied) [inline build.gradle:438]: Need to resolve this comment

**Comment by peternied** [comment]
Note; Moving this to draft while I'm making sure these tests are stable, I've added a bunch of debug only changes that should be reverted before merging, I'll pull this out of draft when its mergeable - but feel free to review it as it stands

**Comment by peternied** [comment]
With CrossCluster tests disabled additional tests to look into:
 - [ ] org.opensearch.security.SearchOperationTest.shouldUpdateTemplate_positive
 - [ ] org.opensearch.security.SearchOperationTest.shouldCreateIndexTemplate_positive

... waiting on additional results, will edit this comment as they come in from the latest [[run]](https://github.com/opensearch-project/security/actions/runs/6278834267/job/17053392199?pr=3388)

**Comment by peternied** [comment]
OK - I'm pretty sure this is ready to merge, or near as ready as I'm going to try to push it, I've made a number of annotations to ignore flaky tests, we should revisit these and I'll add several tracking issues, along with disabling windows due to the network trouble.

Proof that this test collection is stable - 10x runs on JDK 11 / 17
- https://github.com/opensearch-project/security/actions/runs/6305277168/job/17118384937?pr=3388

**Review by willyborankin — APPROVED** [review]
LGTM

**Comment by peternied** [comment]
@RyanL1997 @DarshitChanpura could I get another review?

**Review by DarshitChanpura — APPROVED** [review]
LGTM. I left one question out of curiosity.

Nit: There are some new files being introduced with floraGunn license headers (there is one more license header as well in a couple of files). We should clean this up at some point. Not necessarily in this PR, as these need to be fixed in main as well.

**Review by scrawfor99 — APPROVED** [review]

File: build.gradle
Comment (RyanL1997) [reply build.gradle:438]: Should we remove this btw? Or is there any reason for forcing this? I do see that we have transitive dependencies here: https://github.com/opensearch-project/security/blob/ba3a7014d1c9c19e07f5087742bcf4ab59e04a88/build.gradle#L494

**Review by RyanL1997 — APPROVED** [review]
Hi @peternied, thanks for taking this on and in generally this is looking good to me. I just left a minor comment for the `httpcomponents` dependencies.

File: build.gradle
Comment (peternied) [reply build.gradle:478]: I disabled running these tests on windows due to the failures - it seems like hostname verification might have been causing issues, but this patch didn't work.  I'll remove this post merge in a follow up PR

File: build.gradle
Comment (peternied) [reply build.gradle:438]: I'll update it in a fast follow PR

**Comment by peternied** [comment]
@DarshitChanpura & @RyanL1997 see [1] for the fast follow up items.
- [1] https://github.com/opensearch-project/security/pull/3427

**Pull request merged**
